redis/lookup.go

```diff
@@ -0,0 +1,177 @@
+	if frequency == 0 {
+		// Use default frequency.
+		frequency = 30 * time.Second
+	}
+	if dnsAuthority != "" {
+		host, port, err := net.SplitHostPort(dnsAuthority)
+		if err != nil {
+			// Assume only hostname or IPv4 address was specified.
+			host = dnsAuthority
+			port = "53"
+		}
+		dnsAuthority = net.JoinHostPort(host, port)
+	}
+	return &Lookup{
+		srvLookups:      srvLookups,
+		updateFrequency: frequency,
+		dnsAuthority:    dnsAuthority,
+		ring:            ring,
+		logger:          logger,
+	}
+}
+
+// getResolver returns a resolver that will be used to perform SRV lookups.
+func (look *Lookup) getResolver() *net.Resolver {
+	if look.dnsAuthority == "" {
+		return net.DefaultResolver
+	}
+	return &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			return net.Dial(network, look.dnsAuthority)
+		},
+	}
+}
+
+// handleDNSError logs non-temporary DNS errors and returns nil. Temporary DNS
+// errors are returned as-is.
+func (look *Lookup) handleDNSError(err error, lookupType string, srv cmd.ServiceDomain) error {
+	if err != nil {
+		return nil
+	}
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) && (dnsErr.IsTimeout || dnsErr.IsTemporary) {
+		return err
+	}
+	// Non-temporary DNS errors should always be logged as they are indicative
+	// of a misconfiguration.
+	look.logger.Errf("resolving redis shards, %s lookup for %+v failed: %s", lookupType, srv, err)
+	return nil
+}
+
+// Shards performs SRV lookups for the given service name and returns the
+// resolved shard addresses. An error is only returned if all lookups fail
+// and/or 0 shards are resolved.
+func (look *Lookup) Shards(ctx context.Context) (map[string]string, error) {
+	resolver := look.getResolver()
+
+	var tempErrs []error
+	newAddrs := make(map[string]string)
+	for _, srv := range look.srvLookups {
+		_, targets, err := resolver.LookupSRV(ctx, srv.Service, "tcp", srv.Domain)
+		err = look.handleDNSError(err, "SRV", srv)
+		if err != nil {
+			tempErrs = append(tempErrs, err)
+			// Skip to the next SRV lookup.
+			continue
+		}
+
+		for _, target := range targets {
+			host := strings.TrimRight(target.Target, ".")
+			if look.dnsAuthority != "" {
+				// Lookup A/AAAA records for the SRV target using the custom DNS
+				// authority.
+				hostAddrs, err := resolver.LookupHost(ctx, host)
+				err = look.handleDNSError(err, "A/AAAA", srv)
+				if err != nil {
+					tempErrs = append(tempErrs, err)
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				if len(hostAddrs) == 0 {
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				// Use the first resolved IP address.
+				host = hostAddrs[0]
+			}
+			addr := fmt.Sprintf("%s:%d", host, target.Port)
+			newAddrs[addr] = addr
+		}
+	}
+	// Only return an error if all lookups failed.
+	if len(tempErrs) > 0 && len(newAddrs) == 0 {
+		return nil, errors.Join(tempErrs...)
+	}
+	return newAddrs, nil
```

Comment:
What if `len(newAddrs) == 0` but there were no errors, e.g. because `len(look.srvLookups) == 0`? Seems like that's a time to return an error of our own creation, or a reason to assert that `len(look.srvLookups) > 0` in `NewLookup()`.

Also this would allow you to remove the extra `if len(addrs) == 0` checks from both shardsPeriodically and Start, so there's less duplication of code between the sync and async calls to look.Shards.